File: compress_string.py
def compress_string(s: str) -> str:
    if not s:
        return ""

    compressed = []
    count = 1

    for i in range(1, len(s)):
        if s[i] == s[i - 1]:
            count += 1
        else:
            compressed.append(s[i - 1])
            compressed.append(str(count))
            count = 1

    # Append the last character and its count
    compressed.append(s[-1])
    compressed.append(str(count))

    compressed_string = ''.join(compressed)
    
    # Only return the compressed string if it's shorter than the original
    return compressed_string if len(compressed_string) < len(s) else s

# Test cases
print(compress_string("aabcccccaaa"))  # Expected output: a2b1c5a3
print(compress_string("abc"))          # Expected output: abc (since compression isn't beneficial)
print(compress_string(""))             # Expected output: "" (empty string)
print(compress_string("aabbcc"))       # Expected output: aabbcc (since compression isn't beneficial)
print(compress_string("aaAAaa"))       # Expected output: aaAAaa (since compression isn't beneficial)

File: compress_string.js
function compressString(s) {
    if (!s) {
        return "";
    }

    let compressed = [];
    let count = 1;

    for (let i = 1; i < s.length; i++) {
        if (s[i] === s[i - 1]) {
            count++;
        } else {
            compressed.push(s[i - 1]);
            compressed.push(count.toString());
            count = 1;
        }
    }

    // Append the last character and its count
    compressed.push(s[s.length - 1]);
    compressed.push(count.toString());

    let compressedString = compressed.join('');

    // Only return the compressed string if it's shorter than the original
    return compressedString.length < s.length ? compressedString : s;
}

// Test cases
console.log(compressString("aabcccccaaa"));  // Expected output: a2b1c5a3
console.log(compressString("abc"));          // Expected output: abc (since compression isn't beneficial)
console.log(compressString(""));             // Expected output: "" (empty string)
console.log(compressString("aabbcc"));       // Expected output: aabbcc (since compression isn't beneficial)
console.log(compressString("aaAAaa"));       // Expected output: aaAAaa (since compression isn't beneficial)
